### PR TITLE
Extend i8g sizes, while adding i8ge, i7i, & i7ie SKUs

### PR DESCRIPTION
**Pull request by serprex**
> [!IMPORTANT]
> Adds support for AWS instance types `i8ge`, `i7i`, and `i7ie`, updating options, feature flags, and CLI documentation.
> 
>   - **Behavior**:
>     - Adds support for AWS instance types `i8ge`, `i7i`, and `i7ie` in `helpers/postgres.rb` and `lib/option.rb`.
>     - Updates `generate_postgres_options()` in `helpers/postgres.rb` to include new instance types.
>   - **Options**:
>     - Updates `AWS_FAMILY_OPTIONS` and `AWS_STORAGE_SIZE_OPTIONS` in `lib/option.rb` to include `i8ge`, `i7i`, and `i7ie`.
>     - Adds new size options for `i8ge`, `i7i`, and `i7ie` in `POSTGRES_SIZE_OPTIONS` in `lib/option.rb`.
>   - **Feature Flags**:
>     - Adds feature flags `enable_i7i`, `enable_i7ie`, and `enable_i8ge` in `model/project.rb`.
>   - **CLI Documentation**:
>     - Updates CLI help files in `spec/routes/api/cli/golden-files/` to include new instance types in allowed option values.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ubicloud%2Fubicloud&utm_source=github&utm_medium=referral)<sup> for afed01db695bd1216534c29aad6f07c880475433. You can [customize](https://app.ellipsis.dev/ubicloud/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: lib/option.rb [hunk 1]
```diff
@@ -48,7 +48,10 @@ def self.vring_workers(vcpus)
     "m7a" => non_storage_optimized_vm_storage_size_options,
     "m7i" => non_storage_optimized_vm_storage_size_options,
     "m8gd" => non_storage_optimized_vm_storage_size_options,
-    "i8g" => {2 => ["468"], 4 => ["937"], 8 => ["1875"], 16 => ["3750"], 32 => ["7500"], 64 => ["15000"]}
+    "i8g" => {2 => ["468"], 4 => ["937"], 8 => ["1875"], 16 => ["3750"], 32 => ["7500"], 64 => ["15000"], 96 => ["22500"]},
+    "i8ge" => {2 => ["1250"], 4 => ["2500"], 8 => ["5000"], 12 => ["7500"], 24 => ["15000"], 48 => ["30000"], 72 => ["45000"], 96 => ["60000"]},
+    "i7i" => {2 => ["468"], 4 => ["937"], 8 => ["1875"], 16 => ["3750"], 32 => ["7500"], 48 => ["11250"], 64 => ["15000"], 96 => ["22500"]},
+    "i7ie" => {2 => ["468"], 4 => ["937"], 8 => ["1875"], 12 => ["2812"], 24 => ["5625"], 48 => ["11250"], 72 => ["16875"], 96 => ["22500"]}
```
Comment (iamKunalGupta) [inline lib/option.rb]:
```suggestion
    "i7ie" => {2 => ["1250"], 4 => ["2500"], 8 => ["5000"], 12 => ["7500"], 24 => ["15000"], 48 => ["30000"], 72 => ["45000"], 96 => ["60000"]}
```

**Comment by serprex** [comment]
add config billing with rates at 0

**Comment by serprex** [comment]
blocked on #4325 since adding tests hit edge case unit tests had missed

File: config/billing_rates.yml [hunk 2]
```diff
@@ -186,36 +201,66 @@
...
-- { id: 8d7f1082-2020-4e08-845a-8ade2f14a23a, resource_type: VmVCpu,                      resource_family: i8g,                        location: us-east-2,      unit_price: 0.0000000000, billed_by: duration, active_from: 2025-12-01T00:00:00Z, byoc: true }
-- { id: 0e854a86-b52f-4dfa-8ed3-c3f8832ff3f7, resource_type: VmVCpu,                      resource_family: i8g,                        location: ap-southeast-2, unit_price: 0.0000000000, billed_by: duration, active_from: 2025-12-01T00:00:00Z, byoc: true }
-- { id: 5d4448ed-ae65-44a4-9238-fad1913c85f7, resource_type: VmVCpu,                      resource_family: i8g,                        location: eu-west-1,      unit_price: 0.0000000000, billed_by: duration, active_from: 2025-12-01T00:00:00Z, byoc: true }
+- { id: 44c6a0a4-5f1b-4465-a6ef-647e1b4e2535, resource_type: VmVCpu,                      resource_family: i8g,                        location: us-west-2,      unit_price: 0.0000000000, billed_by: duration, active_from: 2025-12-01T00:00:00Z, byoc: true }
```
Comment (byucesoy) [inline config/billing_rates.yml]: Why do we change the ids of i8g billing rates?

File: helpers/postgres.rb [hunk 3]
```diff
@@ -127,7 +127,7 @@ def generate_postgres_options(flavor: nil, location: nil)
 
     options.add_option(name: "family", values: Option::POSTGRES_FAMILY_OPTIONS.keys, parent: "location") do |flavor, location, family|
       if location.aws?
-        ["m8gd", "i8g"].include?(family) || (Option::AWS_FAMILY_OPTIONS.include?(family) && @project.send(:"get_ff_enable_#{family}"))
+        ["m8gd", "i8g", "i8ge", "i7i", "i7ie"].include?(family) || (Option::AWS_FAMILY_OPTIONS.include?(family) && @project.send(:"get_ff_enable_#{family}"))
```
Comment (byucesoy) [inline helpers/postgres.rb]: The left part of the `||` condition are for families that don't require feature flag. I think we would want these to be behind feature flag.

File: config/billing_rates.yml [hunk 4]
```diff
@@ -285,6 +345,36 @@
 - { id: ab65600f-912c-45cb-b3fa-e26351b1f91c, resource_type: PostgresStandbyVCpu,         resource_family: standard-i8g,               location: us-east-2,      unit_price: 0.0024854760, billed_by: duration, active_from: 2025-09-15T00:00:00Z, byoc: false }
 - { id: 4af9014c-ee3f-45df-99bc-ec489600d846, resource_type: PostgresStandbyVCpu,         resource_family: standard-i8g,               location: ap-southeast-2, unit_price: 0.0029793850, billed_by: duration, active_from: 2025-09-15T00:00:00Z, byoc: false }
 - { id: 273aa797-bcb1-4d71-9151-ed4a1e7d2152, resource_type: PostgresStandbyVCpu,         resource_family: standard-i8g,               location: eu-west-1,      unit_price: 0.0027389480, billed_by: duration, active_from: 2025-09-15T00:00:00Z, byoc: false }
+- { id: 399e03bc-580f-473c-b7b5-0fba746c700d, resource_type: PostgresVCpu,                resource_family: standard-i8ge,              location: us-west-2,      unit_price: 0.0024854760, billed_by: duration, active_from: 2025-09-15T00:00:00Z, byoc: false }
```
Comment (byucesoy) [inline config/billing_rates.yml]: How did you calculated appropriate values for these billing rates in the case of `byoc: false` and `byoc: true`?

File: prog/vm/nexus.rb [hunk 5]
```diff
@@ -125,7 +125,15 @@ def self.assemble(public_key, project_id, name: nil, size: DEFAULT_SIZE,
       prog = if location.aws?
         disk_index = 0
         storage_volumes.each do |volume|
-          max_disk_size = (vm_size.family == "i8g") ? 3750.0 : 1900.0
+          max_disk_size =
+            case vm_size.family
+            when "i8ge", "i7ie"
+              7500.0
```
Comment (byucesoy) [inline prog/vm/nexus.rb]: It seems `2xlarge` for both of these families doesn't fit into the pattern. It comes with 2 * 2500 GB disks.

image

You can either add a special condition here or we might switch to detecting device counts & sizes from inside of the VM with something like lsblk. Latter would be preferable but it is more work.

Comment (byucesoy) [reply prog/vm/nexus.rb]: [This](https://instances.vantage.sh/?id=8a00a011ad189e63b7bc24f5f33da589dbeda017) is the page I checked this information BTW

It is sad that 3rd party website provides better information than AWS itself.

File: prog/vm/nexus.rb [hunk 6]
```diff
@@ -12,7 +12,7 @@ def self.assemble(public_key, project_id, name: nil, size: DEFAULT_SIZE,
     distinct_storage_devices: false, force_host_id: nil, exclude_host_ids: [], gpu_count: 0, gpu_device: nil,
     hugepages: true, hypervisor: nil, ch_version: nil, firmware_version: nil, new_private_subnet_name: nil,
     exclude_availability_zones: [], availability_zone: nil, alternative_families: [],
-    allow_private_subnet_in_other_project: false, init_script: nil)
+    allow_private_subnet_in_other_project: false, init_script: nil, skip_billing: false)
```
Comment (byucesoy) [inline prog/vm/nexus.rb]: I guess this and the change below is not required anymore, right?

File: config/billing_rates.yml (hunk 2 above)
Comment (serprex) [reply config/billing_rates.yml]: mistake, will revert those lines

File: config/billing_rates.yml (hunk 4 above)
Comment (serprex) [reply config/billing_rates.yml]: copied i8g, will zero

**Comment by furkansahin** [comment]
Please update the commit message according to https://github.com/ubicloud/ubicloud/blob/main/COMMIT_MESSAGES.md

**Comment by serprex** [comment]
I assume the qualm was with "aws:" prefix, as I don't think there's really anything more to put in body about this PR

**Comment by serprex** [comment]
@jeremyevans while verbose, it's good to have this information accessible. maybe as a separate command `aws list-sizes` or something. But nailing down how best to go about this seems best as a separate PR

**Comment by serprex** [comment]
@byucesoy suggests filtering aws SKU lists to what's enabled for project

**Comment by jeremyevans** [comment]
> @byucesoy suggests filtering aws SKU lists to what's enabled for project

That will require some work in Rodish, as currently, help output is static per-command, not dependent on request-specific context.

**Comment by jeremyevans** [comment]
> > @byucesoy suggests filtering aws SKU lists to what's enabled for project
> 
> That will require some work in Rodish, as currently, help output is static per-command, not dependent on request-specific context.

I did some research on this, and Rodish already supports context sensitive help, though we aren't currently using it. Enabling it would be something like:

```diff
diff --git a/cli-commands/help.rb b/cli-commands/help.rb
index 5c5b0799e..f3b5c808a 100644
--- a/cli-commands/help.rb
+++ b/cli-commands/help.rb
@@ -29,9 +29,7 @@ UbiCli.on("help") do
               body << banner << "\n"
             end
           else
-            # When we want to switch to context-sensitive help
-            # body << cmd.context_help(self) << "\n\n"
-            body << cmd.help << "\n\n"
+            body << cmd.context_help(self) << "\n\n"
           end
         end
         response(body)
@@ -42,9 +40,7 @@ UbiCli.on("help") do
         end
         response(body)
       else
-        # When we want to switch to context-sensitive help
-        # response(command.context_help(self))
-        response(command.help)
+        response(command.context_help(self))
       end
     else
       orig_command.raise_failure("invalid command: #{argv.join(" ")}")
diff --git a/cli-commands/vm/post/create.rb b/cli-commands/vm/post/create.rb
index a23df49d0..c16573769 100644
--- a/cli-commands/vm/post/create.rb
+++ b/cli-commands/vm/post/create.rb
@@ -17,7 +17,7 @@ UbiCli.on("vm").run_on("create") do
     on("-u", "--unix-user=username", "username (default: ubi)")
   end
   help_option_values("Boot Image:", Option::BootImages.map(&:name))
-  help_option_values("Size:", server_sizes)
+  help_option_values("Size:") { project_specific_vm_sizes }
   help_option_values("Storage Size:", storage_sizes)
 
   help_example 'ubi vm eu-central-h1/my-vm-name create "$(cat ~/.ssh/id_ed25519.pub)"'
diff --git a/lib/ubi_cli.rb b/lib/ubi_cli.rb
index 775c91637..be5641a57 100644
--- a/lib/ubi_cli.rb
+++ b/lib/ubi_cli.rb
@@ -77,9 +77,12 @@ class UbiCli
     [status, {"content-type" => "text/plain", "content-length" => message.bytesize.to_s}, [message]]
   rescue Rodish::CommandFailure => e
     status = 400
-    message = e.message_with_usage.dup
+    message = e.message.dup
     message[0] = "! #{message[0].capitalize}"
-    message += "\n" unless message.end_with?("\n")
+    if e.command
+      message << "\n\n" << e.command.context_help(new(env))
+    end
+    message << "\n" unless message.end_with?("\n")
 
     [status, {"content-type" => "text/plain", "content-length" => message.bytesize.to_s}, [message]]
   end
```

The issue here is the `project_specific_vm_sizes` code. This runs in CLI context, so it cannot access the database directly. We would first have to add an API endpoint that returns available VM sizes for the project. @byucesoy discussed how the possible endpoint should look in  https://github.com/ubicloud/ubicloud/pull/4257#discussion_r2565090942 . If it was implemented, that could be used to find available sizes. I worry that the endpoint return result would be quite large, though.

**Comment by serprex** [comment]
can those improvements be left out of scope for this PR?

**Comment by jeremyevans** [comment]
> can those improvements be left out of scope for this PR?

Yes.

**Review by furkansahin — APPROVED** [review]
